File: kd-tree/Version1/ChatGPT/kd-tree.c
// kd_tree.c  —  C port of kd_tree.py  (Barnes–Hut-style k-d tree)
// Converted from: kd_tree.py  (θ = 0.3, MAX_PARTS = 7)  — see citation in chat.
//
// Build (single-threaded):
//   cc -O3 -march=native -flto kd_tree.c -o kd_tree -lm
//
// Build (OpenMP parallel accel loop):
//   cc -O3 -march=native -flto -fopenmp kd_tree.c -o kd_tree -lm
//
// Run demo:
//   ./kd_tree  (generates a small random system, runs a few steps)

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <float.h>
#include <time.h>

#ifdef _OPENMP
  #include <omp.h>
#endif

// ----- Constants (match the Python) -----
#ifndef MAX_PARTS
#define MAX_PARTS 7
#endif

#ifndef THETA
#define THETA 0.3
#endif

// Optional softening to avoid singularities on exact overlap in point–point accel.
// (The Python pairwise function wasn't provided; this keeps things robust.)
#ifndef SOFTENING2
#define SOFTENING2 0.0
#endif

// ===== Math / types ==========================================================
typedef struct { double x, y, z; } Vec3;

static inline Vec3 v3(double x, double y, double z) { Vec3 r = {x,y,z}; return r; }
static inline Vec3 v3_add(Vec3 a, Vec3 b){ return v3(a.x+b.x, a.y+b.y, a.z+b.z); }
static inline Vec3 v3_sub(Vec3 a, Vec3 b){ return v3(a.x-b.x, a.y-b.y, a.z-b.z); }
static inline Vec3 v3_scale(Vec3 a, double s){ return v3(a.x*s, a.y*s, a.z*s); }
static inline double v3_dot(Vec3 a, Vec3 b){ return a.x*b.x + a.y*b.y + a.z*b.z; }

// ===== Particle ==============================================================
typedef struct {
    Vec3 p;   // position
    Vec3 v;   // velocity
    double m; // mass
} Particle;

// Pairwise acceleration on 'a' due to 'b':  a <- a - (m_b * (a - b) / |a-b|^3)
static inline Vec3 calc_pp_accel(const Particle* a, const Particle* b) {
    Vec3 dp = v3_sub(a->p, b->p);
    double r2 = v3_dot(dp, dp) + SOFTENING2;
    double r = sqrt(r2);
    if (r == 0.0) return v3(0,0,0);
    double mag = -b->m / (r2 * r);
    return v3_scale(dp, mag);
}

// ===== KD tree node =========================================================
typedef struct {
    // Leaf
    int   num_parts;   // >0 => leaf; ==0 => internal
    int*  particles;   // length = num_parts (indices into particle array)

    // Internal
    int   split_dim;   // 0/1/2
    double split_val;
    double m;          // total mass in node
    Vec3  cm;          // center of mass
    double size;       // extent along split_dim (max - min)
    int   left;        // index of left child node in nodes[]
    int   right;       // index of right child node in nodes[]
} KDNode;

typedef struct {
    int     n;           // number of bodies
    int*    indices;     // permutation of [0..n)
    KDNode* nodes;       // contiguous node storage
    int     nodes_cap;   // capacity
} System;

// ---- Node helpers ----
static void node_clear(KDNode* nd){
    if (nd->particles){ free(nd->particles); }
    memset(nd, 0, sizeof(KDNode));
}
static KDNode node_make_leaf(int count, const int* idx_src){
    KDNode nd; memset(&nd, 0, sizeof(nd));
    nd.num_parts = count;
    nd.particles = (int*)malloc(sizeof(int)*count);
    for(int i=0;i<count;i++) nd.particles[i] = idx_src[i];
    return nd;
}
static void ensure_nodes_capacity(System* sys, int need_index){
    if (need_index < sys->nodes_cap) return;
    int newcap = sys->nodes_cap ? sys->nodes_cap : 2;
    while (newcap <= need_index) newcap = newcap*2 + 16;
    sys->nodes = (KDNode*)realloc(sys->nodes, sizeof(KDNode)*newcap);
    for (int i = sys->nodes_cap; i < newcap; ++i){
        sys->nodes[i].particles = NULL;
        sys->nodes[i].num_parts = 0;
        sys->nodes[i].split_dim = 0;
        sys->nodes[i].split_val = 0.0;
        sys->nodes[i].m = 0.0;
        sys->nodes[i].cm = v3(0,0,0);
        sys->nodes[i].size = 0.0;
        sys->nodes[i].left = sys->nodes[i].right = 0;
    }
    sys->nodes_cap = newcap;
}

// ---- System init/free -------------------------------------------------------
static void system_init(System* sys, int n){
    sys->n = n;
    sys->indices = (int*)malloc(sizeof(int)*n);
    for (int i=0;i<n;i++) sys->indices[i] = i;
    // initial guess for nodes, from Python: 2 * (n // (MAX_PARTS-1) + 1)
    int guess = 2 * (n / (MAX_PARTS-1) + 1);
    sys->nodes = (KDNode*)calloc(guess, sizeof(KDNode));
    sys->nodes_cap = guess;
}
static void system_clear_nodes(System* sys){
    for (int i=0;i<sys->nodes_cap;i++){
        if (sys->nodes[i].num_parts > 0 && sys->nodes[i].particles){
            free(sys->nodes[i].particles);
            sys->nodes[i].particles = NULL;
        }
        sys->nodes[i].num_parts = 0;
        // zeroing other fields not strictly required each step
    }
}
static void system_free(System* sys){
    if (sys->nodes){
        for (int i=0;i<sys->nodes_cap;i++){
            if (sys->nodes[i].particles) free(sys->nodes[i].particles);
        }
        free(sys->nodes);
    }
    free(sys->indices);
    memset(sys, 0, sizeof(*sys));
}

// ---- Partition/quickselect (as in the Python loop) --------------------------
static inline double coord(const Particle* p, int dim){
    return (dim==0)?p->p.x : (dim==1)?p->p.y : p->p.z;
}

// Partition-based nth element to put the k-th element (start+k) in place by split_dim
static void nth_element_indices(int* idx, int start, int end, int k, const Particle* parts, int split_dim){
    // We reproduce the randomized loop style used in the Python code.
    // Invariant: we keep shrinking [s,e) until pivot lands at mid (start+k).
    int s = start, e = end;
    int mid = start + k;
    while (s + 1 < e){
        int pivot = s + rand() % (e - s);

        // swap idx[s], idx[pivot]
        int tmp = idx[s]; idx[s] = idx[pivot]; idx[pivot] = tmp;

        int low = s + 1;
        int high = e - 1;
        while (low <= high){
            if (coord(&parts[idx[low]], split_dim) < coord(&parts[idx[s]], split_dim)){
                low++;
            } else {
                int t = idx[low]; idx[low] = idx[high]; idx[high] = t;
                high--;
            }
        }
        // place pivot to its final position at 'high'
        int t = idx[s]; idx[s] = idx[high]; idx[high] = t;

        if (high < mid){
            s = high + 1;
        } else if (high > mid){
            e = high;
        } else {
            // done
            return;
        }
    }
    // trivial tail: if only one element remains, it's already correct
}

// ---- Build tree -------------------------------------------------------------
static int build_tree(System* sys, int start, int end, const Particle* parts, int cur_node){
    int np1 = end - start;
    if (np1 <= MAX_PARTS){
        ensure_nodes_capacity(sys, cur_node);
        // clear any previous leaf storage before overwriting
        if (sys->nodes[cur_node].num_parts > 0 && sys->nodes[cur_node].particles){
            free(sys->nodes[cur_node].particles);
            sys->nodes[cur_node].particles = NULL;
        }
        sys->nodes[cur_node] = node_make_leaf(np1, &sys->indices[start]);
        return cur_node;
    } else {
        // bounding box + mass/CM
        Vec3 minv = v3( 1e100,  1e100,  1e100);
        Vec3 maxv = v3(-1e100, -1e100, -1e100);
        double m = 0.0;
        Vec3 cm = v3(0,0,0);

        for (int i=start;i<end;i++){
            const Particle* p = &parts[ sys->indices[i] ];
            m += p->m;
            cm = v3_add(cm, v3_scale(p->p, p->m));
            if (p->p.x < minv.x) minv.x = p->p.x;
            if (p->p.y < minv.y) minv.y = p->p.y;
            if (p->p.z < minv.z) minv.z = p->p.z;
            if (p->p.x > maxv.x) maxv.x = p->p.x;
            if (p->p.y > maxv.y) maxv.y = p->p.y;
            if (p->p.z > maxv.z) maxv.z = p->p.z;
        }
        cm = v3_scale(cm, 1.0 / m);

        // choose split dim with largest span (the Python had a FIXME; we do the natural 0..2)
        double span[3] = { maxv.x - minv.x, maxv.y - minv.y, maxv.z - minv.z };
        int split_dim = 0;
        if (span[1] > span[split_dim]) split_dim = 1;
        if (span[2] > span[split_dim]) split_dim = 2;
        double size = span[split_dim];

        // partition around median on split_dim
        int mid = (start + end) / 2;
        nth_element_indices(sys->indices, start, end, mid - start, parts, split_dim);
        double split_val = coord(&parts[ sys->indices[mid] ], split_dim);

        // recurse
        int left_root = cur_node + 1;
        int left_last = build_tree(sys, start, mid, parts, left_root);
        int right_root = left_last + 1;
        int right_last = build_tree(sys, mid, end, parts, right_root);

        ensure_nodes_capacity(sys, cur_node);
        // clear any previous leaf storage
        if (sys->nodes[cur_node].num_parts > 0 && sys->nodes[cur_node].particles){
            free(sys->nodes[cur_node].particles);
            sys->nodes[cur_node].particles = NULL;
        }
        sys->nodes[cur_node].num_parts = 0; // mark internal
        sys->nodes[cur_node].split_dim = split_dim;
        sys->nodes[cur_node].split_val = split_val;
        sys->nodes[cur_node].m = m;
        sys->nodes[cur_node].cm = cm;
        sys->nodes[cur_node].size = size;
        sys->nodes[cur_node].left = left_root;
        sys->nodes[cur_node].right = right_root;
        return right_last;
    }
}

// ---- Acceleration recursion -------------------------------------------------
static Vec3 accel_recur(int cur_node, int p_index, const Particle* parts, const KDNode* nodes){
    const KDNode* node = &nodes[cur_node];
    if (node->num_parts > 0){
        Vec3 acc = v3(0,0,0);
        for (int i=0;i<node->num_parts;i++){
            int q = node->particles[i];
            if (q != p_index){
                Vec3 a = calc_pp_accel(&parts[p_index], &parts[q]);
                acc = v3_add(acc, a);
            }
        }
        return acc;
    } else {
        Vec3 dp = v3_sub(parts[p_index].p, node->cm);
        double dist2 = v3_dot(dp, dp);
        // Barnes–Hut criterion: size^2 < θ^2 * r^2
        if (node->size * node->size < (THETA * THETA) * dist2){
            double r = sqrt(dist2);
            if (r == 0.0) return v3(0,0,0);
            double mag = -node->m / (dist2 * r);
            return v3_scale(dp, mag);
        } else {
            Vec3 aL = accel_recur(node->left,  p_index, parts, nodes);
            Vec3 aR = accel_recur(node->right, p_index, parts, nodes);
            return v3_add(aL, aR);
        }
    }
}
static inline Vec3 calc_accel(int p, const Particle* parts, const KDNode* nodes){
    return accel_recur(0, p, parts, nodes);
}

// ---- Utility: print tree (compatible with the Python format) ---------------
static void print_tree_file(int step, const KDNode* nodes, int nodes_cap, const Particle* parts, int n){
    char fname[64]; snprintf(fname, sizeof(fname), "tree%d.txt", step);
    FILE* f = fopen(fname, "w");
    if (!f) return;
    fprintf(f, "%d\n", n);
    for (int i=0;i<nodes_cap;i++){
        const KDNode* nd = &nodes[i];
        if (nd->num_parts > 0){
            fprintf(f, "L %d\n", nd->num_parts);
            for (int j=0;j<nd->num_parts;j++){
                int p = nd->particles[j];
                fprintf(f, "%.17g %.17g %.17g\n", parts[p].p.x, parts[p].p.y, parts[p].p.z);
            }
        } else {
            fprintf(f, "I %d %.17g %d %d\n", nd->split_dim, nd->split_val, nd->left, nd->right);
        }
    }
    fclose(f);
}

// ---- Simple integrator (like simple_sim in Python) -------------------------
static void simple_sim(Particle* bodies, int n, double dt, int steps, int print_steps){
    System sys; system_init(&sys, n);
    Vec3* acc = (Vec3*)malloc(sizeof(Vec3)*n);

    for (int step=0; step<steps; ++step){
        if (print_steps) printf("%d\n", step);

        for (int i=0;i<n;i++) sys.indices[i] = i;
        system_clear_nodes(&sys);
        build_tree(&sys, 0, n, bodies, 0);

        // per-body acceleration (parallelizable)
        #pragma omp parallel for schedule(static)
        for (int i=0;i<n;i++){
            acc[i] = calc_accel(i, bodies, sys.nodes);
        }

        // integrate (kick-drift Euler, matching Python’s simple update)
        for (int i=0;i<n;i++){
            bodies[i].v = v3_add(bodies[i].v, v3_scale(acc[i], dt));
            bodies[i].p = v3_add(bodies[i].p, v3_scale(bodies[i].v, dt));
        }

        // Optionally dump the tree for debugging/visualization:
        // if (step % 10 == 0) print_tree_file(step, sys.nodes, sys.nodes_cap, bodies, n);
    }

    free(acc);
    system_free(&sys);
}

// ===== Demo main =============================================================
#ifdef DEMO_MAIN
int main(void){
    srand((unsigned)time(NULL));

    int n = 20000;            // adjust as desired
    double dt = 1e-3;
    int steps = 5;

    Particle* bodies = (Particle*)malloc(sizeof(Particle)*n);
    // random ring-like cloud
    for (int i=0;i<n;i++){
        double r = 50.0 + 10.0 * ((double)rand()/RAND_MAX - 0.5);
        double ang = 2.0*M_PI*((double)rand()/RAND_MAX);
        double z = 2.0 * ((double)rand()/RAND_MAX - 0.5);
        bodies[i].p = v3(r*cos(ang), r*sin(ang), z);
        bodies[i].v = v3(0.0, 0.0, 0.0);
        bodies[i].m = 1.0/n;
    }

    simple_sim(bodies, n, dt, steps, 1);

    // print final COM just to produce a tiny output
    Vec3 cm = v3(0,0,0); double m=0.0;
    for (int i=0;i<n;i++){ cm = v3_add(cm, v3_scale(bodies[i].p, bodies[i].m)); m += bodies[i].m; }
    cm = v3_scale(cm, 1.0/m);
    printf("Final CM ~ (%.6f, %.6f, %.6f)\n", cm.x, cm.y, cm.z);

    free(bodies);
    return 0;
}
#else
// Provide a no-op main for easy compile if DEMO_MAIN not set.
int main(void){
    fprintf(stderr, "Built kd_tree library. Define DEMO_MAIN to run a demo.\n");
    return 0;
}
#endif

File: kd-tree/Version1/ChatGPT/kd-tree.rs
use rand::prelude::*;
use rayon::prelude::*;
use std::cmp::Ordering;
use std::f64::consts::PI;

// ===== parameters (match the Python) =========================================
const MAX_PARTS: usize = 7;   // leaf threshold
const THETA: f64 = 0.3;       // Barnes–Hut opening angle
const SOFTENING2: f64 = 0.0;  // set >0 if you want Plummer softening

// ===== basic math =============================================================
#[derive(Clone, Copy, Debug, Default)]
struct Vec3 { x: f64, y: f64, z: f64 }
impl Vec3 {
    #[inline] fn new(x: f64, y: f64, z: f64) -> Self { Self { x, y, z } }
    #[inline] fn add(self, o: Self) -> Self { Self::new(self.x + o.x, self.y + o.y, self.z + o.z) }
    #[inline] fn sub(self, o: Self) -> Self { Self::new(self.x - o.x, self.y - o.y, self.z - o.z) }
    #[inline] fn scale(self, s: f64) -> Self { Self::new(self.x * s, self.y * s, self.z * s) }
    #[inline] fn dot(self, o: Self) -> f64 { self.x*o.x + self.y*o.y + self.z*o.z }
}

// ===== particle ===============================================================
#[derive(Clone, Debug)]
struct Particle {
    p: Vec3,  // position
    v: Vec3,  // velocity
    m: f64,   // mass
}

#[inline]
fn coord(p: &Particle, dim: usize) -> f64 {
    match dim {
        0 => p.p.x,
        1 => p.p.y,
        _ => p.p.z,
    }
}

// Pairwise accel on a due to b: a <- a - m_b*(r_vec)/|r|^3
#[inline]
fn accel_pair(a: &Particle, b: &Particle) -> Vec3 {
    let dp = a.p.sub(b.p);
    let r2 = dp.dot(dp) + SOFTENING2;
    if r2 == 0.0 { return Vec3::new(0.0, 0.0, 0.0); }
    let r = r2.sqrt();
    let mag = -b.m / (r2 * r);
    dp.scale(mag)
}

// ===== KD node ================================================================
#[derive(Clone, Debug)]
struct KDNode {
    // leaf: Some(particle_indices); internal: None
    leaf_particles: Option<Vec<usize>>,

    // internal data (valid when leaf_particles is None)
    split_dim: usize,
    split_val: f64,
    m: f64,
    cm: Vec3,
    size: f64,        // extent along split_dim
    left: usize,      // index of left child in nodes vec
    right: usize,     // index of right child
}
impl KDNode {
    fn new_leaf(idxs: &[usize]) -> Self {
        Self {
            leaf_particles: Some(idxs.to_vec()),
            split_dim: 0, split_val: 0.0, m: 0.0, cm: Vec3::default(),
            size: 0.0, left: 0, right: 0,
        }
    }
    fn new_internal(split_dim: usize, split_val: f64, m: f64, cm: Vec3, size: f64, left: usize, right: usize) -> Self {
        Self {
            leaf_particles: None,
            split_dim, split_val, m, cm, size, left, right,
        }
    }
    #[inline]
    fn is_leaf(&self) -> bool { self.leaf_particles.is_some() }
}

// ===== Quickselect nth_element over index slice by coordinate ================
fn nth_element_indices(
    idx: &mut [usize],
    parts: &[Particle],
    split_dim: usize,
    n_th: usize, // 0-based inside idx-slice
) {
    let mut rng = thread_rng();
    let mut left = 0usize;
    let mut right = idx.len();

    while left + 1 < right {
        let pivot_i = left + rng.gen_range(0..(right - left));
        idx.swap(left, pivot_i);

        // partition around pivot at left
        let pv = coord(&parts[idx[left]], split_dim);
        let mut lo = left + 1;
        let mut hi = right - 1;

        while lo <= hi {
            if coord(&parts[idx[lo]], split_dim) < pv {
                lo += 1;
            } else {
                idx.swap(lo, hi);
                if hi == 0 { break; }
                hi -= 1;
            }
        }
        idx.swap(left, hi);

        match hi.cmp(&(left + n_th)) {
            Ordering::Less => left = hi + 1,
            Ordering::Greater => right = hi,
            Ordering::Equal => return,
        }
    }
}

// ===== Tree building =========================================================
struct KDTree {
    nodes: Vec<KDNode>,
    indices: Vec<usize>, // permutation of 0..n
}

impl KDTree {
    fn build(parts: &[Particle]) -> Self {
        let n = parts.len();
        let mut tree = Self { nodes: Vec::with_capacity(2 * (n / (MAX_PARTS.saturating_sub(1).max(1))) + 16),
                              indices: (0..n).collect() };
        // reserve a root slot
        tree.rec_build(parts, 0, n, 0);
        tree
    }

    // returns last node index created (continuous layout)
    fn rec_build(&mut self, parts: &[Particle], start: usize, end: usize, cur_node: usize) -> usize {
        if self.nodes.len() <= cur_node {
            self.nodes.resize_with(cur_node + 1, || KDNode::new_leaf(&[]));
        }

        let count = end - start;
        if count <= MAX_PARTS {
            // leaf
            let leaf = KDNode::new_leaf(&self.indices[start..end]);
            self.nodes[cur_node] = leaf;
            return cur_node;
        }

        // bounding box + mass/CM
        let mut minv = Vec3::new( f64::INFINITY,  f64::INFINITY,  f64::INFINITY);
        let mut maxv = Vec3::new(-f64::INFINITY, -f64::INFINITY, -f64::INFINITY);
        let mut m = 0.0;
        let mut cm = Vec3::new(0.0, 0.0, 0.0);
        for &i in &self.indices[start..end] {
            let p = &parts[i];
            m += p.m;
            cm = cm.add(p.p.scale(p.m));
            if p.p.x < minv.x { minv.x = p.p.x; }
            if p.p.y < minv.y { minv.y = p.p.y; }
            if p.p.z < minv.z { minv.z = p.p.z; }
            if p.p.x > maxv.x { maxv.x = p.p.x; }
            if p.p.y > maxv.y { maxv.y = p.p.y; }
            if p.p.z > maxv.z { maxv.z = p.p.z; }
        }
        cm = cm.scale(1.0 / m);

        let span = [maxv.x - minv.x, maxv.y - minv.y, maxv.z - minv.z];
        let mut split_dim = 0usize;
        if span[1] > span[split_dim] { split_dim = 1; }
        if span[2] > span[split_dim] { split_dim = 2; }
        let size = span[split_dim];

        // partition indices around median on split_dim
        let mid = (start + end) / 2;
        nth_element_indices(&mut self.indices[start..end], parts, split_dim, mid - start);
        let split_val = coord(&parts[self.indices[mid]], split_dim);

        // recurse: left is next slot, right follows left subtree
        let left_root = cur_node + 1;
        let left_last = self.rec_build(parts, start, mid, left_root);
        let right_root = left_last + 1;
        let right_last = self.rec_build(parts, mid, end, right_root);

        self.nodes[cur_node] = KDNode::new_internal(split_dim, split_val, m, cm, size, left_root, right_root);
        right_last
    }

    // acceleration at particle i
    fn accel(&self, i: usize, parts: &[Particle]) -> Vec3 {
        self.accel_recur(0, i, parts)
    }

    fn accel_recur(&self, node_idx: usize, i: usize, parts: &[Particle]) -> Vec3 {
        let nd = &self.nodes[node_idx];
        if let Some(ref ids) = nd.leaf_particles {
            let mut a = Vec3::new(0.0, 0.0, 0.0);
            for &q in ids {
                if q != i {
                    a = a.add(accel_pair(&parts[i], &parts[q]));
                }
            }
            a
        } else {
            let dp = parts[i].p.sub(nd.cm);
            let r2 = dp.dot(dp);
            if nd.size * nd.size < (THETA * THETA) * r2 {
                // use COM approximation
                let r = r2.sqrt();
                if r == 0.0 { return Vec3::new(0.0, 0.0, 0.0); }
                let mag = -nd.m / (r2 * r);
                dp.scale(mag)
            } else {
                let a_l = self.accel_recur(nd.left,  i, parts);
                let a_r = self.accel_recur(nd.right, i, parts);
                a_l.add(a_r)
            }
        }
    }
}

// ===== Simple stepper (like your simple_sim) =================================
fn step_system(parts: &mut [Particle], dt: f64) {
    let tree = KDTree::build(parts);

    // Parallel per-body accelerations
    let accels: Vec<Vec3> = (0..parts.len())
        .into_par_iter()
        .map(|i| tree.accel(i, parts))
        .collect();

    // Kick-drift Euler (matches typical simple loop)
    for (p, a) in parts.iter_mut().zip(accels.into_iter()) {
        p.v = p.v.add(a.scale(dt));
        p.p = p.p.add(p.v.scale(dt));
    }
}

// ===== Demo main =============================================================
fn main() {
    // Small demo; adjust N/steps to taste
    let n = 20_000usize;
    let steps = 5usize;
    let dt = 1e-3_f64;

    let mut rng = StdRng::seed_from_u64(42);
    let mut bodies: Vec<Particle> = (0..n).map(|_| {
        // Put points in a noisy ring for something nontrivial
        let r = 50.0 + 10.0 * (rng.gen::<f64>() - 0.5);
        let ang = 2.0 * PI * rng.gen::<f64>();
        let z = 2.0 * (rng.gen::<f64>() - 0.5);
        Particle {
            p: Vec3::new(r * ang.cos(), r * ang.sin(), z),
            v: Vec3::new(0.0, 0.0, 0.0),
            m: 1.0 / n as f64,
        }
    }).collect();

    for s in 0..steps {
        println!("step {}", s);
        step_system(&mut bodies, dt);
    }

    // Print final COM as a tiny check
    let (mut cm, mut msum) = (Vec3::new(0.0,0.0,0.0), 0.0);
    for p in &bodies { cm = cm.add(p.p.scale(p.m)); msum += p.m; }
    cm = cm.scale(1.0 / msum);
    println!("final COM ≈ ({:.6}, {:.6}, {:.6})", cm.x, cm.y, cm.z);
}
